tmt/steps/__init__.py

```diff
@@ -828,12 +876,32 @@ def _patch_raw_datum(
 
             self.plan._applied_cli_invocations.append(invocation)
 
+        cli_invocations = self.__class__.cli_invocations
+
+        from tmt.steps import discover, execute, finish, prepare, provision, report
+
+        # Under some conditions, the step may need to have at least one phase
+        # defined.
+        if isinstance(self, (discover.Discover, execute.Execute, report.Report)):
+            if self.enabled:
+                raw_data = _ensure_default_phase()
+
+        elif isinstance(self, provision.Provision):
+            if self.enabled or cli_invocations and \
+                    any(invocation.options['how'] for invocation in cli_invocations):
+                raw_data = _ensure_default_phase()
```

Comment:
Doesn't `any provision invocation contains how` condition imply that the `provision` command was provided on the command line and thus the step must be enabled? In other words, shouldn't this be simplified as `if self.enabled`?

Reply:
Sounds plausible, although I expect it to blow up - if it wouldn't be ruining a test somewhere, I wouldn't end up making`cli_invocations` check part of the condition ;)

I'll give it a try because I can't imagine (or recall) how a `provision` step can have CLI invocation and not be enabled. The `self.enabled` runs the following:

```python
    def enabled(self) -> Optional[bool]:
        """ True if the step is enabled """
        if self.plan.my_run is None or self.plan.my_run._cli_context_object is None:
            return None

        return self.name in self.plan.my_run._cli_context_object.steps
```

That looks like you could be right, on the other hand, it's a game of whac-a-mole and I would be very, very surprised if no tests hit this harmless iceberg.

Reply:
And the winner is, `tmt try` :)

Namely `/tests/try/basic`:

```
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
::   Default Plan
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

:: [ 11:05:40 ] :: [  BEGIN   ] :: Running 'TMT_CONFIG_DIR=/tmp/tmp.iZXubjJN2N ./local.exp'
spawn tmt try fedora@local
/var/tmp/tmt/run-053
Let's try /tests/base/bad, /tests/base/good, /tests/base/weird and 3 more tests with /default/plan on fedora@local.

/default/plan
    discover
        how: fmf
        order: 50
        directory: /home/happz/git/tmt/tests/try/basic/data
        summary: 6 tests selected
            /tests/base/bad
            /tests/base/good
            /tests/base/weird
            /tests/core/bad
            /tests/core/good
            /tests/core/weird
    provision
    
        summary: 0 guests provisioned
    prepare
    
        summary: 0 preparations applied
    execute

ESC[31mNo guests available for execution.ESC[0m
send: spawn id exp4 not open
    while executing
"send -- "q\r""
    (file "./local.exp" line 6)
:: [ 11:05:41 ] :: [   FAIL   ] :: Command 'TMT_CONFIG_DIR=/tmp/tmp.iZXubjJN2N ./local.exp' (Expected 0, got 1)
:: [ 11:05:41 ] :: [   PASS   ] :: File '/var/tmp/rlRun_LOG.BX2fu54m' should contain 'Let's try.*/default/plan' 
:: [ 11:05:41 ] :: [   FAIL   ] :: File '/var/tmp/rlRun_LOG.BX2fu54m' should contain 'Run .* successfully finished. Bye for now!' 
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
::   Duration: 1s
::   Assertions: 1 good, 2 bad
::   RESULT: FAIL (Default Plan)
```

The `provision` step is **not** enabled, but there is a custom CLI invocation, `CliInvocation(context=None, options={'image': 'fedora', 'how': 'local'})`. Hence the `or` & test for CLI invocations. `self.name in self.plan.my_run._cli_context_object.steps` fails because `steps` is an empty set.

@psss what would you propose? Shall `try` modify the `steps` set, to mark steps as enabled?